File: app/src/main/java/com/nuhkoca/myapplication/helper/AppExecutors.java
package com.nuhkoca.myapplication.helper;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import javax.inject.Inject;

import androidx.annotation.NonNull;

/**
 * A class that makes I/O, background or UI based processes.
 *
 * @author nuhkoca
 */
public class AppExecutors {
    private final Executor networkIO;
    private final Executor diskIO;
    private final Executor mainIO;

    @Inject
    AppExecutors() {
        this(Executors.newSingleThreadExecutor(), Executors.newFixedThreadPool(Constants.EXECUTOR_THREAD_POOL_OFFSET));
    }

    /**
     * Default constructor
     *
     * @param diskIO    represents background threads
     * @param networkIO represents network threads
     */
    private AppExecutors(Executor diskIO, Executor networkIO) {
        this.diskIO = diskIO;
        this.networkIO = networkIO;
        this.mainIO = new MainThreadExecutor();
    }

    /**
     * A class for {@link #mainIO}
     */
    static class MainThreadExecutor implements Executor {
        private Handler mainThreadHandler = new Handler(Looper.getMainLooper());

        @Override
        public void execute(@NonNull Runnable command) {
            mainThreadHandler.post(command);
        }
    }

    /**
     * Returns network executor
     *
     * @return {@link Executor}
     */
    public Executor networkIO() {
        return networkIO;
    }

    /**
     * Returns I/O executor
     *
     * @return {@link Executor}
     */
    public Executor diskIO() {
        return diskIO;
    }

    /**
     * Return UI executor
     *
     * @return {@link Executor}
     */
    public Executor mainIO() {
        return mainIO;
    }
}

File: app/src/main/java/com/nuhkoca/myapplication/helper/AppExecutors.kt
package com.nuhkoca.myapplication.helper

import android.os.Handler
import android.os.Looper

import java.util.concurrent.Executor
import java.util.concurrent.Executors

import javax.inject.Inject

/**
 * A class that makes I/O, background or UI based processes.
 *
 * @param diskIO    represents background thread
 * @param networkIO represents network threads
 *
 * @author nuhkoca
 */
class AppExecutors @Inject constructor() {
    private val mainIO: Executor
    private val diskIO: Executor
    private val networkIO: Executor

    init {
        this.mainIO = MainThreadExecutor()
        this.diskIO = Executors.newSingleThreadExecutor()
        this.networkIO = Executors.newFixedThreadPool(Constants.EXECUTOR_THREAD_POOL_OFFSET)
    }

    /**
     * A class for [mainIO]
     */
    internal class MainThreadExecutor : Executor {
        private val mainThreadHandler = Handler(Looper.getMainLooper())

        override fun execute(command: Runnable) {
            mainThreadHandler.post(command)
        }
    }

    /**
     * Returns network executor
     *
     * @return [Executor]
     */
    fun networkIO(): Executor {
        return networkIO
    }

    /**
     * Returns I/O executor
     *
     * @return [Executor]
     */
    fun diskIO(): Executor {
        return diskIO
    }

    /**
     * Return UI executor
     *
     * @return [Executor]
     */
    fun mainIO(): Executor {
        return mainIO
    }
}
